File: socialblocker/control.py
"""High-level operations shared by the CLI, GUI, and daemon.

Every mutation goes: load state -> change -> save state -> re-apply hosts.
This keeps /etc/hosts in sync with the source-of-truth state file, and enforces
the locked-session rule in exactly one place.
"""

from __future__ import annotations

import subprocess
import time

from . import autostart, config, hosts_engine
from .config import BLACKLIST, WHITELIST, OFF, MODES, Session, ScheduleRule


class LockedError(RuntimeError):
    """Raised when the user tries to weaken protection during a locked session."""


def _guard_locked(state: config.State) -> None:
    if state.session.active() and state.session.locked:
        mins = (state.session.remaining() + 59) // 60
        raise LockedError(
            f"A locked focus session is running. It cannot be stopped for "
            f"another {mins} min. That is the point of locked mode."
        )


def _reap_completed(state: config.State) -> bool:
    """Record a focus session that has ended, then clear it. Returns True if it
    changed `state`.

    A session that runs to completion has no natural event to hook, so every
    load path reaps here (and so does the daemon). Recording is deduped on the
    session's `ends_at`, which makes it idempotent when the daemon and a CLI
    command race to reap the same expiry — the minutes are only counted once.
    """
    s = state.session
    if s.ends_at == 0 or s.active():
        return False  # nothing running, or still running
    # A boot block is enforcement, not focus work. Logging it would report ~300
    # "focused" minutes every single day, which makes the streak meaningless.
    already = (s.source == config.BOOT
               or any(abs(r.ended_at - s.ends_at) < 1.0 for r in state.focus_log))
    if not already:
        minutes = round((s.ends_at - s.started_at) / 60) if s.started_at else 0
        state.focus_log.append(config.FocusRecord(
            ended_at=s.ends_at, minutes=max(0, minutes),
            mode=s.mode, locked=s.locked))
        state.prune_log()
    state.session = Session()  # cleared; the session is over
    return True


def _load() -> config.State:
    """Load state and reap any just-completed session in-memory. Callers that
    save afterwards persist the reap; read-only callers use `reap()` instead."""
    state = config.load_state()
    _reap_completed(state)
    return state


def reap() -> bool:
    """Persistently log a just-completed session, if there is one. Used by the
    daemon (and status()) so stats stay current even without a mutating op."""
    state = config.load_state()
    if _reap_completed(state):
        config.save_state(state)
        return True
    return False


def refresh() -> tuple[str, bool, int]:
    """Re-apply whatever the state says should be enforced right now."""
    state = config.load_state()
    if _reap_completed(state):
        config.save_state(state)
    return hosts_engine.apply(state)


# --- default (all-day) mode ------------------------------------------------

def set_default_mode(mode: str) -> tuple[str, bool, int]:
    if mode not in MODES:
        raise ValueError(f"mode must be one of {MODES}")
    state = _load()
    # Weakening the default mode during a locked session is not allowed.
    if mode == OFF or (mode == BLACKLIST and state.default_mode == WHITELIST):
        _guard_locked(state)
    state.default_mode = mode
    config.save_state(state)
    return hosts_engine.apply(state)


# --- focus sessions --------------------------------------------------------

def start_session(minutes: int, mode: str = WHITELIST,
                  locked: bool = False) -> tuple[str, bool, int]:
    if mode not in (BLACKLIST, WHITELIST):
        raise ValueError("session mode must be 'blacklist' or 'whitelist'")
    if minutes <= 0:
        raise ValueError("session length must be positive")
    state = _load()
    if state.session.active():
        _guard_locked(state)  # cannot replace a locked session
    now = time.time()
    state.session = Session(
        mode=mode,
        started_at=now,
        ends_at=now + minutes * 60,
        locked=locked,
    )
    config.save_state(state)
    return hosts_engine.apply(state)


def stop_session() -> tuple[str, bool, int]:
    state = _load()
    _guard_locked(state)
    state.session = Session()  # cleared (stopping early does not count as focus)
    config.save_state(state)
    return hosts_engine.apply(state)


def extend_session(minutes: int) -> tuple[str, bool, int]:
    """Extending is always allowed — even during a locked session."""
    state = _load()
    if not state.session.active():
        raise RuntimeError("no active session to extend")
    state.session.ends_at += minutes * 60
    config.save_state(state)
    return hosts_engine.apply(state)


# --- list editing ----------------------------------------------------------

def _norm(domain: str) -> str:
    return domain.strip().lower().lstrip(".")


def add_to_blocklist(domains: list[str]) -> None:
    state = _load()
    state.blocklist = sorted(set(state.blocklist) | {_norm(d) for d in domains if _norm(d)})
    config.save_state(state)
    hosts_engine.apply(state)


def remove_from_blocklist(domains: list[str]) -> None:
    state = _load()
    # Shrinking protection during a locked blacklist session is blocked.
    if state.session.active() and state.session.mode == BLACKLIST:
        _guard_locked(state)
    drop = {_norm(d) for d in domains}
    state.blocklist = sorted(d for d in state.blocklist if d not in drop)
    config.save_state(state)
    hosts_engine.apply(state)


def add_to_whitelist(domains: list[str]) -> None:
    """Allowing more sites is always fine, even mid-lock (it only loosens the
    strict whitelist a little — the discussion notes this friction is expected)."""
    state = _load()
    state.whitelist = sorted(set(state.whitelist) | {_norm(d) for d in domains if _norm(d)})
    config.save_state(state)
    hosts_engine.apply(state)


def remove_from_whitelist(domains: list[str]) -> None:
    state = _load()
    drop = {_norm(d) for d in domains}
    state.whitelist = sorted(d for d in state.whitelist if d not in drop)
    config.save_state(state)
    hosts_engine.apply(state)


# --- schedules -------------------------------------------------------------

def add_schedule(rule: ScheduleRule) -> None:
    state = _load()
    state.schedules.append(rule)
    config.save_state(state)
    hosts_engine.apply(state)


def clear_schedules() -> None:
    state = _load()
    _guard_locked(state)
    state.schedules = []
    config.save_state(state)
    hosts_engine.apply(state)


# --- autostart -------------------------------------------------------------
#
# Three independent switches, deliberately separate:
#   1. the state flag below  — arm a block once per boot
#   2. the systemd service   — makes the root daemon run at boot at all
#   3. the .desktop entry    — opens the GUI at login (see autostart.py)
# (1) does nothing without (2), because the daemon is what arms it.

SERVICE_NAME = "socialblocker"


def set_autostart(enabled: bool | None = None, minutes: int | None = None,
                  mode: str | None = None,
                  locked: bool | None = None) -> config.Autostart:
    """Update the boot-block config. Only the arguments passed are changed.

    No locked-mode guard here: this configures the *next* boot and cannot
    shorten or weaken the session running right now.
    """
    state = _load()
    a = state.autostart
    was_enabled = a.enabled

    if mode is not None:
        if mode not in (BLACKLIST, WHITELIST):
            raise ValueError("autostart mode must be 'blacklist' or 'whitelist'")
        a.mode = mode
    if minutes is not None:
        if minutes <= 0:
            raise ValueError("autostart length must be positive")
        a.minutes = minutes
    if locked is not None:
        a.locked = bool(locked)
    if enabled is not None:
        a.enabled = bool(enabled)

    # Turning it on stamps the current boot as already handled, so the setting
    # takes effect from the *next* boot. Without this, a daemon crash-restart
    # later today would arm a block the user never asked for right now.
    if a.enabled and not was_enabled:
        a.last_boot_id = config.boot_id()

    config.save_state(state)
    return a


def arm_boot_session() -> hosts_engine.Applied | None:
    """Start the configured boot block, at most once per machine boot.

    Called by the daemon at startup. Returns what was applied, or None when
    nothing was armed. Skipped when:
      - autostart is off, or the boot identity is unknown (fail safe: never
        arm rather than risk arming repeatedly);
      - this boot was already handled (the `Restart=always` case);
      - a session is already running — including a locked one that survived a
        reboot, which must not be shortened or replaced by rebooting.
    """
    state = _load()
    a = state.autostart
    if not a.enabled:
        return None

    bid = config.boot_id()
    if not bid or bid == a.last_boot_id:
        return None
    a.last_boot_id = bid

    if state.session.active():
        config.save_state(state)  # remember this boot; leave the session alone
        return None

    # Written as one state object (key + session) so a single atomic save
    # records both — a crash can't leave the boot marked as handled with no
    # session started. That is why this does not route through start_session().
    now = time.time()
    state.session = Session(
        mode=a.mode,
        started_at=now,
        ends_at=now + a.minutes * 60,
        locked=a.locked,
        source=config.BOOT,
    )
    config.save_state(state)
    return hosts_engine.apply(state)


def _systemctl(*args: str):
    """Run systemctl, or return None if this machine has no systemd."""
    try:
        return subprocess.run(["systemctl"] + list(args),
                              capture_output=True, text=True)
    except OSError:
        return None


def service_enabled() -> bool | None:
    """Is the daemon enabled at boot? None when systemd or the unit is missing.

    Deliberately not part of status(): it spawns a process, and the GUI polls
    status once a second.
    """
    cp = _systemctl("is-enabled", SERVICE_NAME)
    if cp is None:
        return None
    val = (cp.stdout or "").strip()
    if val in ("enabled", "enabled-runtime", "static", "alias", "indirect"):
        return True
    if val in ("disabled", "masked", "masked-runtime"):
        return False
    return None  # "not-found" and friends


def set_service_enabled(on: bool) -> str:
    """Enable (and start) or disable the boot daemon. Needs root."""
    action = ["enable", "--now"] if on else ["disable"]
    cp = _systemctl(*(action + [SERVICE_NAME]))
    if cp is None:
        raise RuntimeError(
            "systemctl not found — this machine does not use systemd, so the "
            "daemon cannot be started at boot automatically.")
    if cp.returncode != 0:
        msg = (cp.stderr or cp.stdout or "").strip()
        raise RuntimeError(
            "systemctl {} {} failed: {}\nIf the unit is missing, install it "
            "with: sudo ./install.sh".format(" ".join(action), SERVICE_NAME, msg))
    # Disable intentionally omits --now: the running daemon keeps enforcing the
    # current rules until reboot, so nothing silently unblocks mid-session.
    return "enabled and started" if on else "disabled (still running until reboot)"


def gui_autostart_enabled() -> bool:
    """Is the login (.desktop) entry installed for the desktop user?"""
    return autostart.enabled()


def set_gui_autostart(on: bool) -> str:
    """Add or remove the login entry. Needs no root — it is the user's own file."""
    if on:
        return "added: {}".format(autostart.enable())
    return "removed" if autostart.disable() else "was not present"


# --- presets (quick-start focus sessions) ----------------------------------

def list_presets() -> list[dict]:
    """The bundled quick-start presets (see data/presets.json)."""
    return config.presets()


def start_preset(name: str, minutes: int | None = None) -> tuple[str, bool, int]:
    """Start a focus session from a named preset.

    A preset is only a starting point: pass `minutes` to override its default
    duration (the UI does this via the minute stepper). Locked-mode enforcement
    still happens in start_session — presets never bypass the choke point.
    """
    for p in config.presets():
        if str(p.get("name", "")).lower() == name.lower():
            mins = minutes if minutes is not None else int(p.get("default_minutes", 25))
            return start_session(mins,
                                 mode=p.get("mode", WHITELIST),
                                 locked=bool(p.get("locked", False)))
    valid = ", ".join(str(p.get("name", "")) for p in config.presets())
    raise ValueError(f"unknown preset '{name}'. Available: {valid}")


# --- status ----------------------------------------------------------------

def status() -> dict:
    state = config.load_state()
    if _reap_completed(state):
        config.save_state(state)
    mode, locked = state.effective()
    # "blocked now" is the truthful effective count from the hosts engine — the
    # same base-domain set it would write, so front-ends never re-derive it.
    blocked_now = len(hosts_engine.domains_for_mode(state, mode))
    return {
        "default_mode": state.default_mode,
        "effective_mode": mode,
        "locked": locked,
        "session_active": state.session.active(),
        "session_remaining": state.session.remaining(),
        "session_mode": state.session.mode,
        "session_source": state.session.source,
        # Elapsed-fraction inputs for a progress display. Derived here so no
        # front-end has to do arithmetic on the session entity, and so `extend`
        # correctly rescales the total rather than overshooting it.
        "session_started_at": state.session.started_at,
        "session_total": max(0, int(state.session.ends_at - state.session.started_at)),
        "blocklist_count": len(state.blocklist),
        "whitelist_count": len(state.whitelist),
        "schedules": len(state.schedules),
        "blocked_now": blocked_now,
        "universe_size": len(config.universe()),
        "focused_today_min": state.focused_today_min(),
        "streak_days": state.streak_days(),
        "autostart_enabled": state.autostart.enabled,
        "autostart_minutes": state.autostart.minutes,
        "autostart_mode": state.autostart.mode,
        "autostart_locked": state.autostart.locked,
        # Which copy of the code answered. With both copy and linked installs
        # possible, this is the only reliable way to tell whether the command
        # you just ran is the checkout you just edited.
        "source_dir": str(config.PROJECT_DIR),
    }

File: android/app/src/main/java/org/socialblocker/core/Control.kt
package org.socialblocker.core

/** Raised when the user tries to weaken protection during a locked session. */
class LockedError(message: String) : RuntimeException(message)

/**
 * What "apply the rules" means to the outside world.
 *
 * The desktop's `control.py` calls the hosts engine directly, because writing a
 * file needs no framework. On a phone the same step has to reach a *running
 * service*, which is an outer-circle thing — so it is injected here instead.
 * That keeps the dependency rule intact: `Control` never imports `vpn`.
 */
interface Enforcer {
    fun apply(rules: RuleSet)
}

/** Everything a front-end needs to render, derived in one place. */
data class Status(
    val defaultMode: String,
    val effectiveMode: String,
    val locked: Boolean,
    val sessionActive: Boolean,
    val sessionRemaining: Int,
    val sessionMode: String,
    val sessionSource: String,
    val sessionStartedAt: Double,
    val sessionTotal: Int,
    val blocklistCount: Int,
    val whitelistCount: Int,
    val schedules: Int,
    val blockedNow: Int,
    val universeSize: Int,
    val focusedTodayMin: Int,
    val streakDays: Int,
    val autostartEnabled: Boolean,
    val autostartMinutes: Int,
    val autostartMode: String,
    val autostartLocked: Boolean,
)

/**
 * High-level operations shared by the UI, the Quick Settings tile and the
 * enforcement service — the phone's `control.py`.
 *
 * Every mutation goes: load state -> change -> save state -> re-apply rules.
 * THE LOCKED-SESSION RULE IS ENFORCED HERE AND NOWHERE ELSE. Do not scatter a
 * lock check into an Activity or the notification handler; add the operation
 * here instead.
 *
 * Every mutating method is `@Synchronized`. The desktop got away without a lock
 * because its CLI, GUI and daemon are separate processes reconciled by an
 * atomic file rename. Here they are threads in one process sharing one file, so
 * two concurrent read-modify-writes would genuinely lose an update — "stop" and
 * "extend" arriving together could resurrect a finished session.
 */
class Control(
    private val store: Store,
    private val engine: BlockEngine,
    private val catalog: Catalog,
    private val enforcer: Enforcer,
) {

    private fun guardLocked(state: State) {
        if (state.session.active() && state.session.locked) {
            val mins = (state.session.remaining() + 59) / 60
            throw LockedError(
                "A locked focus session is running. It cannot be stopped for " +
                    "another $mins min. That is the point of locked mode."
            )
        }
    }

    /**
     * Record a focus session that has ended, then clear it. Returns whether
     * [state] changed.
     *
     * A session that runs to completion has no natural event to hook, so every
     * load path reaps here (and so does the service loop). Recording is deduped
     * on the session's `endsAt`, which makes it idempotent when two callers
     * race to reap the same expiry — the minutes are counted once.
     */
    private fun reapCompleted(state: State): Boolean {
        val s = state.session
        if (s.endsAt == 0.0 || s.active()) return false
        // A boot block is enforcement, not focus work. Logging it would report
        // ~300 "focused" minutes every single day and make the streak a lie.
        val already = s.source == BOOT ||
            state.focusLog.any { Math.abs(it.endsAt - s.endsAt) < 1.0 }
        if (!already) {
            val minutes = if (s.startedAt > 0) Math.round((s.endsAt - s.startedAt) / 60).toInt() else 0
            state.focusLog = state.focusLog + FocusRecord(
                endsAt = s.endsAt, minutes = maxOf(0, minutes),
                mode = s.mode, locked = s.locked,
            )
            state.pruneLog()
        }
        state.session = Session()   // cleared; the session is over
        return true
    }

    /** Load state and reap a just-completed session in memory. Callers that
     *  save afterwards persist the reap; read-only callers use [reap]. */
    private fun load(): State = store.load().also { reapCompleted(it) }

    private fun commit(state: State): RuleSet {
        store.save(state)
        val rules = engine.rules(state)
        enforcer.apply(rules)
        return rules
    }

    /** Persistently log a just-completed session, if there is one. Used by the
     *  service loop so stats stay current even if the UI is never opened. */
    @Synchronized
    fun reap(): Boolean {
        val state = store.load()
        if (reapCompleted(state)) { store.save(state); return true }
        return false
    }

    /** Re-apply whatever the state says should be enforced right now. */
    @Synchronized
    fun refresh(): RuleSet {
        val state = store.load()
        if (reapCompleted(state)) store.save(state)
        val rules = engine.rules(state)
        enforcer.apply(rules)
        return rules
    }

    // --- default (all-day) mode --------------------------------------------

    @Synchronized
    fun setDefaultMode(mode: String): RuleSet {
        require(mode in MODES) { "mode must be one of $MODES" }
        val state = load()
        // Weakening the default mode during a locked session is not allowed.
        if (mode == OFF || (mode == BLACKLIST && state.defaultMode == WHITELIST)) {
            guardLocked(state)
        }
        state.defaultMode = mode
        return commit(state)
    }

    // --- focus sessions -----------------------------------------------------

    @Synchronized
    fun startSession(minutes: Int, mode: String = WHITELIST, locked: Boolean = false): RuleSet {
        require(mode == BLACKLIST || mode == WHITELIST) {
            "session mode must be 'blacklist' or 'whitelist'"
        }
        require(minutes > 0) { "session length must be positive" }
        val state = load()
        if (state.session.active()) guardLocked(state)   // cannot replace a locked session
        val t = now()
        state.session = Session(
            mode = mode, startedAt = t, endsAt = t + minutes * 60, locked = locked,
        )
        return commit(state)
    }

    @Synchronized
    fun stopSession(): RuleSet {
        val state = load()
        guardLocked(state)
        state.session = Session()   // stopping early does not count as focus
        return commit(state)
    }

    /** Extending is always allowed — even during a locked session. */
    @Synchronized
    fun extendSession(minutes: Int): RuleSet {
        val state = load()
        if (!state.session.active()) throw IllegalStateException("no active session to extend")
        state.session = state.session.copy(endsAt = state.session.endsAt + minutes * 60)
        return commit(state)
    }

    // --- list editing -------------------------------------------------------

    @Synchronized
    fun addToBlocklist(domains: List<String>) {
        val state = load()
        val add = domains.map(BlockEngine::normalize).filter { it.isNotEmpty() }
        state.blocklist = (state.blocklist + add).toSortedSet().toList()
        commit(state)
    }

    @Synchronized
    fun removeFromBlocklist(domains: List<String>) {
        val state = load()
        // Shrinking protection during a locked blacklist session is blocked.
        if (state.session.active() && state.session.mode == BLACKLIST) guardLocked(state)
        val drop = domains.map(BlockEngine::normalize).toSet()
        state.blocklist = state.blocklist.filter { BlockEngine.normalize(it) !in drop }
        commit(state)
    }

    /** Allowing more sites is always fine, even mid-lock: it only loosens the
     *  strict whitelist a little, and that friction is expected. */
    @Synchronized
    fun addToWhitelist(domains: List<String>) {
        val state = load()
        val add = domains.map(BlockEngine::normalize).filter { it.isNotEmpty() }
        state.whitelist = (state.whitelist + add).toSortedSet().toList()
        commit(state)
    }

    @Synchronized
    fun removeFromWhitelist(domains: List<String>) {
        val state = load()
        val drop = domains.map(BlockEngine::normalize).toSet()
        state.whitelist = state.whitelist.filter { BlockEngine.normalize(it) !in drop }
        commit(state)
    }

    // --- schedules ----------------------------------------------------------

    @Synchronized
    fun addSchedule(rule: ScheduleRule) {
        val state = load()
        state.schedules = state.schedules + rule
        commit(state)
    }

    @Synchronized
    fun removeSchedule(index: Int) {
        val state = load()
        guardLocked(state)
        state.schedules = state.schedules.filterIndexed { i, _ -> i != index }
        commit(state)
    }

    @Synchronized
    fun clearSchedules() {
        val state = load()
        guardLocked(state)
        state.schedules = emptyList()
        commit(state)
    }

    @Synchronized
    fun schedules(): List<ScheduleRule> = store.load().schedules

    // --- autostart ----------------------------------------------------------

    /**
     * Update the boot-block config. Only the arguments passed are changed.
     *
     * No locked-mode guard: this configures the *next* boot and cannot shorten
     * or weaken the session running right now.
     */
    @Synchronized
    fun setAutostart(
        enabled: Boolean? = null,
        minutes: Int? = null,
        mode: String? = null,
        locked: Boolean? = null,
        bootToken: String = "",
    ): Autostart {
        val state = load()
        var a = state.autostart
        val wasEnabled = a.enabled

        if (mode != null) {
            require(mode == BLACKLIST || mode == WHITELIST) {
                "autostart mode must be 'blacklist' or 'whitelist'"
            }
            a = a.copy(mode = mode)
        }
        if (minutes != null) {
            require(minutes > 0) { "autostart length must be positive" }
            a = a.copy(minutes = minutes)
        }
        if (locked != null) a = a.copy(locked = locked)
        if (enabled != null) a = a.copy(enabled = enabled)

        // Turning it on stamps the current boot as already handled, so the
        // setting takes effect from the *next* boot. Without this, the next
        // time Android revived the service today it would arm a block the user
        // never asked for right now.
        if (a.enabled && !wasEnabled) a = a.copy(lastBootToken = bootToken)

        state.autostart = a
        store.save(state)
        return a
    }

    /**
     * Start the configured boot block, at most once per device boot.
     *
     * Called when BOOT_COMPLETED arrives and again whenever the service starts,
     * because Android gives no guarantee about which happens first. Returns
     * what was armed, or null. Skipped when:
     *  - autostart is off, or the boot identity is unknown (fail safe: never
     *    arm rather than risk arming repeatedly);
     *  - this boot was already handled (the START_STICKY revival case);
     *  - a session is already running — including a locked one that survived a
     *    reboot, which must not be shortened or replaced by rebooting.
     */
    @Synchronized
    fun armBootSession(bootToken: String): RuleSet? {
        val state = load()
        val a = state.autostart
        if (!a.enabled) return null
        if (bootToken.isEmpty() || bootToken == a.lastBootToken) return null

        state.autostart = a.copy(lastBootToken = bootToken)

        if (state.session.active()) {
            store.save(state)   // remember this boot; leave the session alone
            return null
        }

        // Written as one state object (key + session) so a single atomic save
        // records both — a crash cannot leave the boot marked as handled with
        // no session started. That is why this does not route through
        // startSession().
        val t = now()
        state.session = Session(
            mode = a.mode, startedAt = t, endsAt = t + a.minutes * 60,
            locked = a.locked, source = BOOT,
        )
        return commit(state)
    }

    // --- presets ------------------------------------------------------------

    fun listPresets(): List<Preset> = catalog.presets()

    /**
     * Start a focus session from a named preset.
     *
     * A preset is only a starting point: pass [minutes] to override its default
     * duration (the UI does this via the stepper). Locked-mode enforcement
     * still happens in [startSession] — presets never bypass the choke point.
     */
    fun startPreset(name: String, minutes: Int? = null): RuleSet {
        val p = catalog.presets().firstOrNull { it.name.equals(name, ignoreCase = true) }
            ?: throw IllegalArgumentException(
                "unknown preset '$name'. Available: " +
                    catalog.presets().joinToString(", ") { it.name }
            )
        return startSession(minutes ?: p.defaultMinutes, p.mode, p.locked)
    }

    // --- status -------------------------------------------------------------

    @Synchronized
    fun status(): Status {
        val state = store.load()
        if (reapCompleted(state)) store.save(state)
        val (mode, locked) = state.effective()
        // "blocked now" is the truthful effective count from the engine — the
        // same set it would sinkhole, so no front-end re-derives it.
        val blockedNow = engine.domainsForMode(state, mode).size
        return Status(
            defaultMode = state.defaultMode,
            effectiveMode = mode,
            locked = locked,
            sessionActive = state.session.active(),
            sessionRemaining = state.session.remaining(),
            sessionMode = state.session.mode,
            sessionSource = state.session.source,
            // Elapsed-fraction inputs for the progress ring. Derived here so no
            // front-end does arithmetic on the session entity, and so `extend`
            // rescales the total rather than overshooting it.
            sessionStartedAt = state.session.startedAt,
            sessionTotal = maxOf(0, (state.session.endsAt - state.session.startedAt).toInt()),
            blocklistCount = state.blocklist.size,
            whitelistCount = state.whitelist.size,
            schedules = state.schedules.size,
            blockedNow = blockedNow,
            universeSize = catalog.universe().size,
            focusedTodayMin = state.focusedTodayMin(),
            streakDays = state.streakDays(),
            autostartEnabled = state.autostart.enabled,
            autostartMinutes = state.autostart.minutes,
            autostartMode = state.autostart.mode,
            autostartLocked = state.autostart.locked,
        )
    }

    /** Read-only access for front-ends that need the raw lists. */
    fun lists(): Pair<List<String>, List<String>> {
        val s = store.load()
        return s.blocklist to s.whitelist
    }
}
